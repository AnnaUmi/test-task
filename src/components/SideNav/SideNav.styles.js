import styled from 'styled-components';

export const SideNavContainer = styled.div`
  width: 200px;
  min-height: 200px;
  height: fit-content;
  background-color: #939393;
`;

export const SideNavListItemStyled = styled.div`
  padding: 20px 30px;
  font-size: 20px;
  cursor: pointer;
  ${({ isActive }) => (isActive === true ? 'background-color: #7e7e7e; font-weight: bold;' : '')};

  :hover {
    background-color: #bdbdbd;
  }
`;

export const SideNavSubMenuListemStyled = styled.div`
  padding: 20px 60px;
  font-size: 18px;
  cursor: pointer;
  ${({ isActive }) => (isActive === true ? 'background-color: #7e7e7e; font-weight: bold;' : '')};

  :hover {
    background-color: #bdbdbd;
  }
`;

export const SubMenuListItemContainer = styled.div`
  background-color: #d3d3d3;
  ${({ isOpened }) => (isOpened === true ? '' : 'height: 0; overflow: hidden;')};
`;
