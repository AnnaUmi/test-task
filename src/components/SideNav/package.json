{
  "name": "side-nav",
  "version": "0.0.0",
  "private": true,
  "main": "./SideNav.js"
}
