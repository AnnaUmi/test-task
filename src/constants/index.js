import PropTypes from 'prop-types';

/* eslint-disable import/prefer-default-export */
export const NORMALIZE_URL =
  'https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css';

export const SIDE_NAV_ITEM_TYPES = {
  ListItem: 'ListItem',
  SubMenu: 'SubMenu',
};

export const routerProps = PropTypes.shape({
  pathName: PropTypes.string,
  route: PropTypes.string,
  query: PropTypes.object,
  back: PropTypes.func,
  beforePopState: PropTypes.func,
  prefetch: PropTypes.func,
  push: PropTypes.func,
  reload: PropTypes.func,
  replace: PropTypes.func,
});

export const sideNavSubMenuItemProps = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  url: PropTypes.string,
};

export const sideNavItemProps = {
  ...sideNavSubMenuItemProps,
  type: PropTypes.oneOf([SIDE_NAV_ITEM_TYPES.ListItem, SIDE_NAV_ITEM_TYPES.SubMenu]).isRequired,
};

export const noop = () => {};
