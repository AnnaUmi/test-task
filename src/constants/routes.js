import { SIDE_NAV_ITEM_TYPES } from '.';

export const HEADER_NAV_ROUTES = [
  {
    route: '/',
    label: 'PRISM Logo',
    id: '/',
  },
  {
    route: '/brand',
    label: 'Brand',
    id: '/brand',
  },
  {
    route: '/product',
    label: 'Product',
    id: '/product',
  },
  {
    route: '/marketing',
    label: 'Marketing',
    id: '/marketing',
  },
  {
    route: '/components',
    label: 'Components',
    id: '/components',
  },
];

const componentsSideNav = [
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Buttons',
    id: 'components/buttons',
    url: '/components/buttons',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Typography',
    id: 'components/typography',
    url: '/components/typography',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Grid',
    id: 'components/grid',
    url: '/components/grid',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Spacing',
    id: 'components/spacing',
    url: '/components/spacing',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Colors',
    id: 'components/colors',
    url: '/components/colors',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Breadcrumbs',
    id: 'components/breadcrumbs',
    url: '/components/breadcrumbs',
  },
];

export const SIDE_NAV_ROUTES = {
  '/components': componentsSideNav,
};
