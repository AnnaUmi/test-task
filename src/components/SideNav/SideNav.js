import React from 'react';
import PropTypes from 'prop-types';

import SideNavListItem from './SideNavListItem';
import SideNavSubMenu from './SideNavSubMenu';
import { SideNavContainer } from './SideNav.styles';
import { sideNavItemProps, sideNavSubMenuItemProps, SIDE_NAV_ITEM_TYPES } from '../../constants';

const SideNav = ({ menuItems, currRoute }) => (
  <SideNavContainer>
    {menuItems.map(item => {
      let result = null;

      if (item.type === SIDE_NAV_ITEM_TYPES.ListItem) {
        result = <SideNavListItem isActive={currRoute === item.url} item={item} key={item.id} />;
      } else if (item.type === SIDE_NAV_ITEM_TYPES.SubMenu) {
        result = (
          <SideNavSubMenu
            listItems={item.listItems}
            key={item.id}
            label={item.label}
            currRoute={currRoute}
          />
        );
      }

      return result;
    })}
  </SideNavContainer>
);

SideNav.propTypes = {
  menuItems: PropTypes.arrayOf(
    PropTypes.shape({
      ...sideNavItemProps,
      listItems: PropTypes.arrayOf(PropTypes.shape(sideNavSubMenuItemProps)),
    })
  ).isRequired,
  currRoute: PropTypes.string.isRequired,
};

export default SideNav;
