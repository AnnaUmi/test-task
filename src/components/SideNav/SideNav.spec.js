import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { SIDE_NAV_ITEM_TYPES } from '../../constants';
import SideNav from './SideNav';

jest.mock('./SideNavListItem', () => 'SideNavListItem');
jest.mock('./SideNavSubMenu', () => 'SideNavSubMenu');

const menuItems = [
  {
    type: SIDE_NAV_ITEM_TYPES.ListItem,
    label: 'Buttons',
    id: 'components/buttons',
    url: '/components/buttons',
  },
  {
    type: SIDE_NAV_ITEM_TYPES.SubMenu,
    label: 'test2',
    id: 'components/test2',
    listItems: [
      { label: 'subMenu1', id: 'components/subMenu1', url: '/components/test' },
      { label: 'subMenu2', id: 'components/subMenu2', url: '/components/test' },
    ],
  },
];

const invalidList = [
  {
    type: 'invalid',
    label: 'Buttons',
    id: 'components/buttons',
    url: '/components/buttons',
  },
];

describe('SideNav', () => {
  it('should match snapshot when rendering valid item types', () => {
    const tree = renderer.create(<SideNav menuItems={menuItems} currRoute="/components/buttons" />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot when rendering invalid item types', () => {
    const tree = renderer.create(
      <SideNav menuItems={invalidList} currRoute="/components/buttons" />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
