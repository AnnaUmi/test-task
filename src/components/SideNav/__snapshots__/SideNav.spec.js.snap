// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should match snapshot when rendering invalid item types 1`] = `
.c0 {
  width: 200px;
  min-height: 200px;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  background-color: #939393;
}

<div
  className="c0"
/>
`;

exports[`SideNav should match snapshot when rendering valid item types 1`] = `
.c0 {
  width: 200px;
  min-height: 200px;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  background-color: #939393;
}

<div
  className="c0"
>
  <SideNavListItem
    isActive={true}
    item={
      Object {
        "id": "components/buttons",
        "label": "Buttons",
        "type": "ListItem",
        "url": "/components/buttons",
      }
    }
  />
  <SideNavSubMenu
    currRoute="/components/buttons"
    label="test2"
    listItems={
      Array [
        Object {
          "id": "components/subMenu1",
          "label": "subMenu1",
          "url": "/components/test",
        },
        Object {
          "id": "components/subMenu2",
          "label": "subMenu2",
          "url": "/components/test",
        },
      ]
    }
  />
</div>
`;
